client/src/controllers/RulesController.ts

```diff
@@ -0,0 +1,156 @@
+ *
+ */
+export class RulesController extends Controller<HTMLFormElement> {
+  static targets = ['enable'];
+
+  /** Targets will be enabled if the `data-rule` matches the scoped form data, otherwise will be disabled. */
+  declare readonly enableTargets: HTMLElement[];
+  declare readonly hasEnableTarget: boolean;
+
+  declare active: boolean;
+  declare ruleCache: Record<string, Record<string, string[]>>;
+
+  initialize() {
+    this.ruleCache = {};
+    this.resolve = debounce(this.resolve.bind(this), 50);
+  }
+
+  connect() {
+    this.checkTargets();
+  }
+
+  /**
+   * Checks for any targets that will mean that the controller needs to be active.
+   */
+  checkTargets() {
+    this.active = this.hasEnableTarget;
+  }
+
+  /**
+   * Resolve the conditional targets based on the form data and the target(s)
+   * `data-rule` attributes.
+   */
+  resolve() {
+    if (!this.active) return;
+
+    const form = this.element;
+    const formData = Object.fromEntries(new FormData(form).entries());
+
+    [
+      ...this.enableTargets.map((target) => ({ shouldDisable: false, target })),
+    ].forEach(({ shouldDisable, target }) => {
+      this.toggleAttribute(
+        target,
+        this.getIsMatch(formData, this.getRuleData(target))
+          ? !shouldDisable
+          : shouldDisable,
+        'disabled',
+      );
+    });
+
+    this.dispatch('resolved', { bubbles: true, cancelable: false });
+  }
+
+  getIsMatch(
+    formData: Record<string, FormDataEntryValue>,
+    ruleData: Record<string, string[]>,
+  ): boolean {
+    return (
+      ruleData &&
+      Object.entries(ruleData).every(([key, value]) =>
+        value.includes(String(formData[key] || '')),
+      )
+    );
+  }
+
+  getRuleData(target: Element): Record<string, string[]> {
+    if (!target) return {};
+    const ruleStr = target.getAttribute('data-rule');
+    if (!ruleStr) return {};
+
+    // check cache
+    const cachedRule = this.ruleCache[ruleStr];
+    if (cachedRule) return cachedRule;
+
+    // prepare parsed rule data
+    let rule = {};
+
+    if (ruleStr) {
+      try {
+        rule = JSON.parse(ruleStr);
+        if (Array.isArray(rule)) rule = Object.fromEntries(rule);
+      } catch (e) {
+        // Safely ignore JSON parsing errors
+      }
+    }
+
+    // Map through values and convert to array of strings
+    // Allowing falsey values to be treated as an empty string
+    const ruleData = Object.fromEntries(
+      Object.entries(rule).map(([key, value = null]) => [
+        key,
+        castArray(value).map((item) => (item ? String(item) : '')),
+      ]),
+    );
+
+    this.ruleCache[ruleStr] = ruleData;
+
+    return ruleData;
+  }
+
+  toggleAttribute(target, shouldRemove = false, attr = 'hidden') {
+    if (shouldRemove) {
+      target.removeAttribute(attr);
+    } else if (attr === 'disabled') {
+      // eslint-disable-next-line no-param-reassign
+      target.disabled = true;
+    } else {
+      target.setAttribute(attr, attr);
```

Comment:
I don't see any tests nor usage of this with the `hidden` attr, and the only call to this method (in `resolve()`) has the parameter hardcoded to `disabled`. I know you likely meant for this to be reusable with different attributes to be toggled, e.g. `hidden` here to hide the field instead of disabling it. However, I'm not sure this is even possible with the current implementation, and it ends up somewhat muddying the code for what we currently use it for.

For example, we end up adding special cases with "half-measures" here:

I don't understand why we're special-casing `'disabled'` to use `target.disabled = true` instead of `setAttribute("disabled", "")`, yet we use `removeAttribute()` instead of `target.disabled = false`. I'd rather we stick to one approach for both disabling and enabling (either use `setAttribute(attr, "")` and `removeAttribute(attr)`, or `target.disabled = true` and `target.disabled = false`).

Additionally, AFAIK the `hidden` and `disabled` properties behave similarly, as in:

- setting `.disabled = true`/`.hidden = true` will add a `"disabled"`/`"hidden"` attribute with an empty string `""` as the value
  - this is the same as `setAttribute("disabled", "")`/`setAttribute("hidden", "")`, which would also make the respective property return `true`
- setting `.disabled = false`/`.hidden = false` will remove the `"disabled"`/`"hidden"` attribute (if exists)
  - this is the same as `removeAttribute("disabled")`/`removeAttribute("hidden")`, which would also make the respective property return `false`

So, `disabled` doesn't need special casing any more than `hidden` does.

Also, I'm not sure when the `target.setAttribute(attr, attr);` case would be useful (there is no test case that demonstrates this).

Perhaps these are leftovers from the `w-cond` implementation?

Sorry for asking to trim down a controller's implementation yet again, I know we've had similar discussions in the past. However, without a usage in core or a test to verify this works as expected, it's hard to justify the additional complexity here 🙏

Reply:
This is helpful feedback, thankyou. 

I've simplified this whole approach, we can review in the future if we need to abstract out to separate methods.